src/views/ApartmentApp/components/StudentApplication/index.js

```diff
@@ -0,0 +1,794 @@
+   */
+  useEffect(() => {
+    let isSubscribed = true;
+
+    const handleError = (_error) => {
+      // No existing application was found in the database,
+      // or an error occurred while attempting to load the application
+      setApplicationID(null);
+      setEditorUsername((prevEditorUsername) => prevEditorUsername ?? userProfile.AD_Username);
+      setApplicants((prevApplicants) =>
+        prevApplicants.length ? prevApplicants : [{ Profile: userProfile, OffCampusProgram: '' }],
+      );
+      setUnsavedChanges(true);
+    };
+
+    if (!applicationID || !applicationDetails) {
+      setLoading(true);
+      // Check if the current user is on an application. Returns the application ID number if found
+      housing
+        .getCurrentApplicationID()
+        .then((newApplicationID) => {
+          console.log('Retrieved Application ID: ' + newApplicationID); //! DEBUG
+          if (newApplicationID === null || newApplicationID === -1) {
+            // Intentionally trigger the 'catch'
+            throw new Error("Invalid value of 'newApplicationID' = " + newApplicationID);
+          } else if (isSubscribed) {
+            setApplicationID(newApplicationID);
+            setUnsavedChanges(false);
+            housing
+              .getApartmentApplication(newApplicationID)
+              .then((newApplicationDetails) => {
+                if (isSubscribed && newApplicationDetails) {
+                  setApplicationDetails(newApplicationDetails);
+                  setDateSubmitted(newApplicationDetails.DateSubmitted ?? null);
+                  setDateModified(newApplicationDetails.DateModified ?? null);
+                  setEditorUsername(newApplicationDetails.EditorUsername ?? null);
+                  setApplicants((prevApplicants) => {
+                    if (newApplicationDetails.Applicants?.length > 0) {
+                      return newApplicationDetails.Applicants.map(async (newApplicantInfo) => {
+                        if (newApplicantInfo.Profile !== null) {
+                          return newApplicantInfo;
+                        } else {
+                          const newApplicantProfile = await user.getProfileInfo(
+                            newApplicantInfo.Username,
+                          );
+                          return { Profile: newApplicantProfile, ...newApplicantInfo };
+                        }
+                      });
+                    } else {
+                      return prevApplicants;
+                    }
+                  });
+                  setPreferredHalls(newApplicationDetails?.ApartmentChoices ?? []);
+                  setUnsavedChanges(false);
+                }
+              })
+              .catch((error) => {
+                handleError(error);
+              });
+          }
+        })
+        .catch((error) => {
+          handleError(error);
+        });
+      setLoading(false);
+    }
+
+    return () => (isSubscribed = false);
```

Comment:
I believe this is intended to prevent state changes to this component after it unmounts. However, I don't think you realized that it gets run on every re-render, because the previous call to `useEffect` is tossed out and the new one is enabled. That's preventing a lot of your state updates from working.

You should just get rid of this. Updating unmounted components is not actively harmful, and it would only happen if someone leaves the page while the initial load is still taking place, so it should be a rare circumstance.

Reply:
I only added that because there were errors with the thing attempting to make state changes after unmount, and that would cause the page to freeze or crash instead of updating the state when doing something like adding an applicant.

For some reason, the StudentApplication component would unmount every time the user adds an applicant forcing the component to update the array of applicants. It is as if it is unmounting and then mounting a "new" instance or something, but still carrying over the state somehow? I don't fully understand the detailed of this part, all I know was that the page crashed before I added the "isSubscribed" thing and the errors reported that is had attempted to update the state of an unmounted component, and after I added that check it worked correctly during my testing.

I know this probably is a symptom of a larger problem with the setup of this useEffect, we can talk more later